docker/archive/tar.go

```diff
@@ -0,0 +1,210 @@
+package archive
+
+// This implements all of the magical internal tar file handling.
+
+import (
+	"archive/tar"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+// fakeFile allows us to generate tar headers in-memory.
+type fakeFile struct {
+	path string
+	size int64
+}
+
+func (ff *fakeFile) Name() string {
+	return ff.path
+}
+func (ff *fakeFile) Size() int64 {
+	return ff.size
+}
+func (ff *fakeFile) Mode() os.FileMode {
+	return 0444
+}
+func (ff *fakeFile) ModTime() time.Time {
+	return time.Unix(0, 0)
+}
+func (ff *fakeFile) IsDir() bool {
+	return ff.Mode().IsDir()
+}
+func (ff *fakeFile) Sys() interface{} {
+	return nil
+}
+
+func sendTarFile(tw *tar.Writer, path string, size int64, stream io.Reader) error {
+	hdr, err := tar.FileInfoHeader(&fakeFile{
+		path: path,
+		size: size,
+	}, "")
+	if err != nil {
+		return err
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+	actualSize, err := io.Copy(tw, stream)
+	if err != nil {
+		return err
+	}
+	if size != actualSize {
+		return fmt.Errorf("send tar file: %s: mismatch between expected size (%d) and actual size (%d)", path, size, actualSize)
+	}
+	return nil
+}
+
+// tarCache allows for archive read operations to be sped up (and also
+// abstracted). It wraps an io.ReadSeeker with some helpful methods for dealing
+// with tar archives.
+type tarCache struct {
+	file    io.ReadSeeker
+	offsets map[string]int64
+}
+
+// prime regenerates all of the offset information in the tarCache.
+func (tc *tarCache) prime() error {
+	const tarBlockSize = 512
+
+	// Reset everything.
+	tc.offsets = map[string]int64{}
+	if _, err := tc.file.Seek(0, os.SEEK_SET); err != nil {
+		return err
+	}
+
+	// Because of the fact that you can have more than one header for a given
+	// file, it's necessary to store the end offset of the previous entry. This
+	// offset is rounded to the tarBlockSize.
+	var lastEntry int64
+
+	tr := tar.NewReader(tc.file)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		// Add ourselves to the offset list.
+		tc.offsets[hdr.Name] = lastEntry
+
+		// Get current seek(2) offset and figure out where the end of our final
+		// chunk is for the next entry.
+		cur, err := tc.file.Seek(0, os.SEEK_CUR)
+		if err != nil {
+			return err
+		}
+
+		lastEntry = cur + hdr.Size
+
+		// Jump to the next block if this entry didn't round to a block.
+		if (cur+hdr.Size)%512 > 0 {
+			lastEntry += tarBlockSize - (lastEntry % tarBlockSize)
+		}
+	}
```

Comment:
That’s still relying on undocumented implementation properties: There is no API promise that after `tar.Reader.Next()` returns, the underlying stream is positioned at the start of the content of the file; e.g. nothing prohibits `tar.Reader` from buffering input (e.g. reading it one `st_blksize` at a time instead of one tar block at a time).